[stable/dask] Add updated Dask chart

mrocklin opened the pull request:
This creates a new Dask chart that replaces the dask-distributed chart.  In addition to a slimmed down name this uses more modern docker images, supports environment customization through environment variables, links cpu and memory limits to dask cpu and memory limits, ...

See https://github.com/kubernetes/charts/issues/2912

cc @danielfrg

mrocklin commented:
Related changes to docker image here: https://github.com/dask/dask-docker/pull/5

mrocklin commented:
@danielfrg if you have time to review the changes here and in https://github.com/dask/dask-docker/pull/6 I would appreciate it.  

Charts maintainers, I'm curious about how best to go about smoothly renaming a chart.  I would like to rename this from `dask-distributed` to `dask` if possible.

I am also curious about how best to decompose this chart so that it can be used either with or without the Jupyter notebook service.  My current thought is to create a dask-core chart that does not include Jupyter, and then have the Dask chart depend on the dask-core chart.  I am unfamiliar with dependencies though, and so I may be doing something naive here.

I'm running the CLA past the legal department of my company.  I hope for a response back in a day or so.

mrocklin commented:
@k8s-ci-robot I have signed the CLA

danielfrg commented:
Changes look good, I cannot test today but I will try to test tomorrow. Maybe lets merge the docker image one to it gets build on dockerhub first.

unguiculus commented on stable/dask/Chart.yaml:
```diff
@@ -1,10 +1,12 @@
-name: dask-distributed
-version: 2.0.0
+name: dask
+version: 1.0.0
```
What's the reason for this? You rename the chart and decrease the version? Seems weird.

mrocklin replied:
Happy to increase the version if desired.  I decreased the version to 1.0 because of the name change.  I'm happy to adhere to any standard though.

unguiculus commented on stable/dask/values.yaml:
```diff
@@ -44,15 +56,24 @@ worker:
-  image: "jupyter/base-notebook"
-  imageTag: "11be019e4079"
+  image: "daskdev/dask-notebook"
+  imageTag: "latest"
```
Should be a stable version.

mrocklin replied:
Agreed

unguiculus commented:
@kubernetes/charts-maintainers The chart is about to be renamed. Is this something we can/want to handle?

mrocklin commented:
There has been a fair amount of work on this over at https://github.com/dask/helm-chart and things have cleaned up a bit.  If review is restarting let me know and I'll push more up.

unguiculus commented:
@mrocklin Just keep changes coming. It probably doesn't make sense to review now if you push more substantial changes.

mrocklin commented:
cc @jacobtomlinson I'd like to make you aware of this.  After merging current efforts you may want to consider pushing the result upstream to the stable repo here.

unguiculus commented:
FYI: Renaming the chart is not as easy. We'd have to keep the existing one and deprecate it. Think about if you really want this.

mrocklin commented:
Thanks for the comment @unguiculus 

Deprecating and creating a new one still sounds ideal to me.  I may not be aware of maintenance costs of keeping around a deprecated chart though.

prydonius commented:
Once the chart is deprecated there are no maintenance costs. Be aware that Helm doesn't really treat deprecated charts in any special way and e.g. a `helm search dask` will show both charts. When deprecating dask-distributed, we can update the description to begin with "DEPRECATED" so it's more obvious to users.

We need a separate PR to deprecate dask-distributed that:
- sets `deprecated: true` in Chart.yaml
- prepends "DEPRECATED" to the description in Chart.yaml
- Updates the README and NOTES.txt to indicate that it's deprecated and `stable/dask` should be used instead

This PR should be updated to just add the new dask chart and not remove dask-distributed.

@danielfrg are you happy with renaming this?

danielfrg commented:
Yes

mrocklin committed:
Add Dask Chart

This supercedes the previous dask-distributed chart, which will be
deprecated.

This also uses an accurate docker image for the Jupyter server,
and pins docker images generally.

mrocklin commented:
I've updated this PR to add a new `dask` chart.  I'll submit a PR deprecating dask-distributed after (if) this gets merged.  Feedback is generally quite welcome on the structure of the chart here.

mrocklin commented on stable/dask/Chart.yaml:
```diff
@@ -0,0 +1,12 @@
+name: dask
+version: 0.17.0
```
Currently I'm pinning the version of the chart to the version of the main software package.  I'm also happy to go directly to 1.0.0 and follow semver if preferred

unguiculus replied:
Add `appVersion` which should be `0.17.0`. The chart is versioned independently.

unguiculus reviewed (changes requested):
Please read our review guidelines again:
https://github.com/kubenetes/charts/blob/master/REVIEW_GUIDELINES.md

Make sure you configure things as suggested there (e. g. images, resources).

If you use `helm create` with Helm 2.8+ you'll get an initial setup with current best practices applied.

mrocklin committed:
Comment out resource limits

mrocklin committed:
Set chart version to 1.0.0, appVersion to 0.17.0

mrocklin committed:
Update README and chart information

mrocklin committed:
specify nthreads and memory-limit explicitly in values.yaml

unguiculus commented:
For the new chart, please compare with the templates and the `_helpers.tpl` file you get with `helm create` using Helm 2.8+ and update accordingly.

mrocklin committed:
update notes

mrocklin committed:
avoid setting --nthreads and --memory-limit if no resource limits

mrocklin committed:
expand truncation to 63 characters, and trim "."

mrocklin committed:
bump app version number

mrocklin commented:
> For the new chart, please compare with the templates and the _helpers.tpl file you get with helm create using Helm 2.8+ and update accordingly.

Thanks for the pointer.  I've half-completed this by adopting the new truncation and stripping conventions.  However I haven't yet adopted the fullnameOverride or `contains $name .Release.Name` logic.  This chart doesn't currently use the `"name.fullname"` construct.  Should I add this anyway?  Is this important for interoperability with other systems?

mrocklin commented:
The linting checks on circle-ci seem off.  They're testing against the old chart `dask-distributed` (against which this PR was originally targetting) rather than the new chart, `dask`.

unguiculus commented:
The linter is correct. It tested both charts in this PR. The old one didn't have a version bump. You should really only have the new chart in this PR. Why is there a change to the old one as well that is supposed to be deprecated in another PR?

mrocklin committed:
remove change to dask-distributed chart

mrocklin commented:
> The linter is correct. It tested both charts in this PR. The old one didn't have a version bump. You should really only have the new chart in this PR. Why is there a change to the old one as well that is supposed to be deprecated in another PR?

Ah, thank you.  I missed that change in the diff.

mrocklin commented:
I could use direction on [this comment](https://github.com/kubernetes/charts/pull/2914#issuecomment-368266148).  Otherwise all concerns have, I believe, been dealt with.

unguiculus reviewed (changes requested):
* The empty `stable/dask/NOTES.txt` should be removed
* Please check what `helm create` using Helm 2.8+ gives you and update accordingly
* Make sure you follow our review guidelines

mrocklin committed:
remove NOTES.txt

mrocklin committed:
use explicit repository and image tags

mrocklin commented:
> The empty stable/dask/NOTES.txt should be removed

Done

> Please check what helm create using Helm 2.8+ gives you and update accordingly

I believe that I have done this, with the exception of the concern raised in https://github.com/kubernetes/charts/pull/2914#issuecomment-368266148 for which I could use help.  Apparently I've failed at this though.  More concrete feedback would be welcome.  

> Make sure you follow our review guidelines

I'm sorry, I looked through these when you made the same comment earlier and I've just read through them again.  I don't see what I'm doing wrong here.  I must be dense here.  If anyone sees what I'm doing wrong I would love detailed feedback.

unguiculus commented on stable/dask/templates/_helpers.tpl:
```diff
@@ -0,0 +1,31 @@
+{{/* vim: set filetype=mustache: */}}
```
Please update with latest `_helpers.tpl` you get from `helm create` using Helm 2.8+.

unguiculus commented on stable/dask/README.md:
```diff
@@ -0,0 +1,85 @@
+| Parameter                  | Description                        | Default                                                    |
+| -------------------------- | ---------------------------------- | ---------------------------------------------------------- |
+| `scheduler.name`           | Dask master name                   | `dask-scheduler`                                           |
+| `scheduler.image`          | Container image name               | `daskdev/dask`                                             |
```
Update image params.

unguiculus commented on stable/dask/templates/dask-jupyter-deployment.yaml:
```diff
@@ -0,0 +1,47 @@
+apiVersion: extensions/v1beta1
```
`apps/v1beta2`

unguiculus commented on stable/dask/templates/dask-jupyter-config.yaml:
```diff
@@ -0,0 +1,14 @@
+    app: {{ template "dask.name" . }}
+    heritage: {{ .Release.Service | quote }}
+    release: {{ .Release.Name | quote }}
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
```
Once you've updated the `_helpers.tpl`, use `chart: {{ template "dask.chart" . }}`.

unguiculus commented on stable/dask/templates/_helpers.tpl:
```diff
@@ -0,0 +1,31 @@
+{{- define "dask.name" -}}
+{{- default .Chart.Name .Values.nameOverride | trunc 63 | trimSuffix "-" -}}
+{{- end -}}
+
```
Fullname templates should be based on the standard fullname template and append a component suffix.

Is there a reason why you make the components dynamic? I'd just hard-code them as you have them in `values.yaml`.

unguiculus commented on stable/dask/templates/dask-jupyter-deployment.yaml:
```diff
@@ -0,0 +1,47 @@
+          image: 
+            repository: {{ .Values.jupyter.image | quote }}
+            tag: {{ .Values.jupyter.imageTag | quote }}
+            pullPolicy: IfNotPresent
```
Make `pullPolicy` configurable.

unguiculus commented on stable/dask/templates/dask-worker-deployment.yaml:
```diff
@@ -0,0 +1,44 @@
+apiVersion: extensions/v1beta1
```
`apps/v1beta2`

unguiculus commented on stable/dask/templates/dask-scheduler-deployment.yaml:
```diff
@@ -0,0 +1,39 @@
+            pullPolicy: IfNotPresent
+          args: ["dask-scheduler", "--port", "{{ .Values.scheduler.servicePort }}", "--bokeh-port", "{{ .Values.webUI.containerPort }}"]
+          ports:
+            - containerPort: {{ .Values.scheduler.containerPort }}
```
Any reason why you make container ports configurable. Changing them may break things. I'd also give them a name and refer to them by name in the services.

unguiculus commented on stable/dask/templates/dask-scheduler-deployment.yaml:
```diff
@@ -0,0 +1,39 @@
+          ports:
+            - containerPort: {{ .Values.scheduler.containerPort }}
+            - containerPort: {{ .Values.webUI.containerPort }}
+          resources: {{ toYaml .Values.scheduler.resources | indent 12 }}
```
```yaml
          resources: 
{{ toYaml .Values.scheduler.resources | indent 12 }}
```

unguiculus commented on stable/dask/templates/dask-worker-deployment.yaml:
```diff
@@ -0,0 +1,44 @@
+                 "--no-bokeh"]
+          ports:
+            - containerPort: {{ .Values.worker.containerPort }}
+          resources: {{ toYaml .Values.worker.resources | indent 12 }}
```
```yaml
          resources: 
{{ toYaml .Values.scheduler.resources | indent 12 }}
```

unguiculus commented on stable/dask/templates/dask-worker-deployment.yaml:
```diff
@@ -0,0 +1,44 @@
+            repository: {{ .Values.worker.image | quote }}
+            tag: {{ .Values.worker.imageTag | quote }}
+            pullPolicy: IfNotPresent
+          args: ["dask-worker",
```
Just a suggestion for a consistent list style:

```yaml
          args: 
            - dask-worker
            - '{{ template "dask.scheduler-fullname" . }}:{{ .Values.scheduler.servicePort }}'
          {{- if .Values.worker.limits -}}
            - --nthreads
            - {{ default .Values.worker.limits.cpu 1 | quote }}
            - --memory-limit
            - {{ default .Values.worker.limits.memory "4GiB" | quote }}
          {{- end -}}
            - --no-bokeh
```
And don't use the `default` function here. Add the defaults to `values.yaml`.

unguiculus reviewed (changes requested):
Sorry, I shouldn't have blindly repeated that. I was on my way home from work and added the comment from my cell phone.

I did a better review now. I'd like you to also indent the lists consistently, either zero or two spaces. Personally, I'd prefer the latter.

mrocklin commented:
Awesome.  Thank you for your time and for the detailed feedback.

mrocklin committed:
respond to feedback

mrocklin committed:
remove specification of containerPorts

mrocklin committed:
update README

mrocklin commented:
OK, I've resolved recent issues.  I'm having difficulty testing things locally with minikube.  I'm now getting the following:

```
mrocklin@carbon:~/workspace/charts/stable/dask$ helm install .
Error: release jumpy-heron failed: Deployment in version "v1beta2" cannot be handled as a Deployment: v1beta2.Deployment: Spec: v1beta2.DeploymentSpec: Template: v1.PodTemplateSpec: Spec: v1.PodSpec: Containers: []v1.Container: v1.Container: Image: ReadString: expects " or n, parsing 670 ...,"image":{... at {"apiVersion":"apps/v1beta2","kind":"Deployment","metadata":{"labels":{"app":"dask","chart":"dask-1.0.0","component":"jumpy-heron-jupyter-notebook","heritage":"Tiller","release":"jumpy-heron"},"name":"jumpy-heron-dask-jupyter","namespace":"default"},"spec":{"replicas":1,"strategy":{"type":"RollingUpdate"},"template":{"metadata":{"labels":{"app":"dask","component":"jumpy-heron-jupyter-notebook","release":"jumpy-heron"}},"spec":{"containers":[{"env":[{"name":"EXTRA_CONDA_PACKAGES","value":null},{"name":"EXTRA_PIP_PACKAGES","value":null},{"name":"EXTRA_APT_PACKAGES","value":null},{"name":"DASK_SCHEDULER_ADDRESS","value":"jumpy-heron-dask-scheduler:8786"}],"image":{"pullPolicy":"IfNotPresent","repository":"daskdev/dask-notebook","tag":"0.17.1"},"name":"jumpy-heron-dask-jupyter","ports":[{"containerPort":8888}],"resources":{},"volumeMounts":[{"mountPath":"/home/jovyan/.jupyter","name":"config-volume"}]}],"volumes":[{"configMap":{"name":"jumpy-heron-dask-scheduler-config"},"name":"config-volume"}]}}}}
```

I suspect that this has to do with my local environment

```
mrocklin@carbon:~/workspace/charts/stable/dask$ helm version
Client: &version.Version{SemVer:"v2.8.1", GitCommit:"6af75a8fd72e2aa18a2b278cfe5c7a1c5feca7f2", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.8.1", GitCommit:"6af75a8fd72e2aa18a2b278cfe5c7a1c5feca7f2", GitTreeState:"clean"}
mrocklin@carbon:~/workspace/charts/stable/dask$ kubectl version
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.0", GitCommit:"925c127ec6b946659ad0fd596fa959be43f0cc05", GitTreeState:"clean", BuildDate:"2017-12-15T21:07:38Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"", Minor:"", GitVersion:"v1.9.0", GitCommit:"925c127ec6b946659ad0fd596fa959be43f0cc05", GitTreeState:"clean", BuildDate:"2018-01-26T19:04:38Z", GoVersion:"go1.9.1", Compiler:"gc", Platform:"linux/amd64"}
```

But I haven't yet found a way to resolve this.

mrocklin committed:
resolve linting issues in values.yaml

mrocklin commented on stable/dask/values.yaml:
```diff
@@ -0,0 +1,83 @@
+  resources: {}
+  aptPackages: >-
+  condaPackages: >-
+  pipPackages: >-
```
Should we include these?  These get placed into environment variables that trigger certain actions when using the docker images described above.  If someone were to use their own docker image (that didn't inherit from ours) then these options would have no effect.  

We could keep these for convenience or remove them and resolve this through documentation, telling users how to specify environment varibles.

unguiculus replied:
I was wondering about those. Since they are empty by default anyways, you should probably remove them and add a generic way to add env vars.

mrocklin replied:
Done.  However I've run into a problem where I want to combine environment variables defined within the template and those that are user provided.  Naively I want the following:

```yaml
          env:
            - name: DASK_SCHEDULER_ADDRESS
              value: {{ template "dask.fullname" . }}-scheduler:{{ .Values.scheduler.servicePort }}
{{ toYaml .Values.jupyter.env | indent 12 }}
```

But this generates the following linting error:

```
$ helm lint .
==> Linting .
[ERROR] templates/dask-jupyter-deployment.yaml: unable to parse YAML
	error converting YAML to JSON: yaml: line 40: could not find expected ':'

Error: 1 chart(s) linted, 1 chart(s) failed
```

mrocklin replied:
I suspect that there is a clever way to do with with the helm templating language.  Happy to go through those docs again if that's the right approach.

mrocklin commented:
cc @jacobtomlinson who expressed some interest in this

mrocklin committed:
Clarify extra packages in README

mrocklin committed:
replace pip/conda/apt packages with generic env solution

jacobtomlinson commented:
A few thoughts from me:
- I would like to see an ingress which is disabled by default but I could easily enable and configure to expose the dashboard externally.
- I would like to be able to disable the notebook that comes with it as it is not 100% necessary.
- I would love an option for the cluster to be adaptive. This would require RBAC for the scheduler to modify the worker deployment. It would also be good to have a min and max workers option in distributed as we have previously discussed. (I appreciate this may be out of scope for now though).

mrocklin commented:
> I would like to see an ingress which is disabled by default but I could easily enable and configure to expose the dashboard externally.

> I would like to be able to disable the notebook that comes with it as it is not 100% necessary.

I'm on board with both of these but have to admit that my familiarity with Helm is limited enough that it's not clear to me today how to implement them.  Do you have any interest in collaborating on this?  I'm giving you push access to my fork now in case the answer is "yes".

> I would love an option for the cluster to be adaptive. This would require RBAC for the scheduler to modify the worker deployment. It would also be good to have a min and max workers option in distributed as we have previously discussed. (I appreciate this may be out of scope for now though).

I agree with this.  If we want min/max then we'll have to wait a bit for that to land in the dask.distributed library.  Things could be incremental though.  Again, I'm not familiar enough with RBAC to be particularly efficient on this.  If you're at all interested I would encourage your involvement.

jdumars commented:
/ok-to-test

mrocklin commented:
OK, I'm currently a bit stuck on a couple things:

1. Despite linting fine my my local environment doesn't run this well.  I don't understand the error messages it's giving me, nor have I yet been able to find suitable information from web searches.

        mrocklin@carbon:~/workspace/charts/stable/dask$ helm install .
        Error: release jumpy-heron failed: Deployment in version "v1beta2" cannot be handled as a Deployment: v1beta2.Deployment: Spec: v1beta2.DeploymentSpec: Template: v1.PodTemplateSpec: Spec: v1.PodSpec: Containers: []v1.Container: v1.Container: Image: ReadString: expects " or n, parsing 670 ...,"image":{... at {"apiVersion":"apps/v1beta2","kind":"Deployment","metadata":{"labels":{"app":"dask","chart":"dask-1.0.0","component":"jumpy-heron-jupyter-notebook","heritage":"Tiller","release":"jumpy-heron"},"name":"jumpy-heron-dask-jupyter","namespace":"default"},"spec":{"replicas":1,"strategy":{"type":"RollingUpdate"},"template":{"metadata":{"labels":{"app":"dask","component":"jumpy-heron-jupyter-notebook","release":"jumpy-heron"}},"spec":{"containers":[{"env":[{"name":"EXTRA_CONDA_PACKAGES","value":null},{"name":"EXTRA_PIP_PACKAGES","value":null},{"name":"EXTRA_APT_PACKAGES","value":null},{"name":"DASK_SCHEDULER_ADDRESS","value":"jumpy-heron-dask-scheduler:8786"}],"image":{"pullPolicy":"IfNotPresent","repository":"daskdev/dask-notebook","tag":"0.17.1"},"name":"jumpy-heron-dask-jupyter","ports":[{"containerPort":8888}],"resources":{},"volumeMounts":[{"mountPath":"/home/jovyan/.jupyter","name":"config-volume"}]}],"volumes":[{"configMap":{"name":"jumpy-heron-dask-scheduler-config"},"name":"config-volume"}]}}}}

2.  I want to drop in a template containing a map in the midst of an ongoing map

    ```yaml
              env:
                - name: DASK_SCHEDULER_ADDRESS
                   value: {{ template "dask.fullname" . }}-scheduler:{{ .Values.scheduler.servicePort }}
    {{ toYaml .Values.jupyter.env | indent 12 }}
    ```

If anyone has any time I would value advice on either of these particular issues.

There are also the issues raised by @jacobtomlinson .  I would love to find a way to turn off the Jupyter server.  Does anyone have any suggestions or examples of completely turning off a segment of the chart completely?

jacobtomlinson commented:
In terms of turning off segments see the [rbac section](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/master/jupyterhub/templates/hub/rbac.yaml) in zero to jupyterhub as an example. The whole template is wrapped in an if statement that just checks a boolean from the config file.

mrocklin committed:
Add option to turn off Jupyter

Also, combine all jupyter documents into one file

mrocklin commented:
Ah, that seems simple enough.  Done.

mrocklin committed:
add missing {{ end }}

unguiculus commented on stable/dask/templates/dask-jupyter.yaml:
```diff
@@ -0,0 +1,84 @@
+    c = get_config()
+    c.NotebookApp.password = '{{ .Values.jupyter.password }}'
+
+---
```
Please create separate files per resource.

mrocklin replied:
Fixed

mrocklin committed:
split dask-jupyter to multiple files

mrocklin commented:
I remain stuck on these two issues: https://github.com/kubernetes/charts/pull/2914#issuecomment-370036473

I've raised one as a question upstream: https://github.com/kubernetes/helm/issues/3602

If anyone has thoughts on how to resolve these I would welcome advice, pointers to documentation etc..

If people have looked at these and genuinely don't know if there is a solution then that would be good feedback as well. Alternatives would be welcome.

unguiculus commented:
I'll try and have a look.

unguiculus committed:
Fix Yaml issues

unguiculus committed:
Fix values.yaml

mrocklin commented:
*Thank you* for interceding here @unguiculus .  This works flawlessly for me locally.

Looking at your changes, was the issue that calling toYaml on an empty mapping was generating something like `{}` rather than empty lines?

mrocklin commented:
I'm going to take another pass over the README and NOTES.  Is there anything else that's necessary here for a 1.0.0 release?

mrocklin committed:
Update README with instructions for installing custom packages

mrocklin commented:
Just checking in here.  I believe all feedback so far has been handled (except for some of @jacobtomlinson's comments, which I think can probably wait for future enhancements).  Is there anything else I should be doing here to move this forward?

also cc @yuvipanda in case he wants to weigh in.

unguiculus commented on stable/dask/templates/dask-jupyter-config.yaml:
```diff
@@ -0,0 +1,18 @@
+    heritage: {{ .Release.Service | quote }}
+    release: {{ .Release.Name | quote }}
+    chart: {{ template "dask.chart" . }}
+    component: "{{ .Release.Name }}-{{ .Values.jupyter.component }}"
```
Let me ask again why you make components configurable. Also, they should not be tied to a release. I'd just hard-code them: `jupyter`, `worker`

mrocklin replied:
Thanks for the nudge.  Resolved.  My apologies for letting that slip earlier. 

To answer the direct question of why was this here.  My answer is "I don't know".  This PR is a copy-paste-modify job of the current chart in `stable/dask-distributed`, that can hopefully be deprecated once this cleaner version is in.

mrocklin committed:
Remove configurable components

unguiculus commented:
/ok-to-test

unguiculus commented:
/lgtm

mrocklin commented:
Thank you for your patient support everyone.  I'm very glad to see this in.
